src/pudl/harvest/harvest.py

```diff
@@ -0,0 +1,1054 @@
+        >>> sample(df, key=['report_month'])['report_month'].dtype.name
+        'datetime64[ns]'
+    """
+    df_key, df_data = find_sample(df=df, key=key, data=data)
+    if len(df_key) < len(key):
+        return None
+    mapper = {old: new for old, new in zip(df_key, key) if new != old}
+    sdf = df[df_key + df_data].rename(columns=mapper)
+    for k, df_k in zip(key, df_key):
+        _, period = split_period(k)
+        if period:
+            sdf[k] = PERIODS[period](sdf[k])
+    return sdf
+
+
+# ---- Column aggregation ---- #
+
+"""
+Aggregation functions.
+
+All take a :class:`pd.Series` as input (and any optional keyword arguments).
+They may either return a single value (ideally of the same data type as the input),
+null (`np.nan`),
+or raise a :class:`ValueError` if the input does not meet requirements.
+"""
+
+
+def most_frequent(x: pd.Series) -> Any:
+    """Return most frequent value (or error if none exists)."""
+    mode = x.mode(dropna=True)
+    if mode.size == 1:
+        return mode[0]
+    if mode.empty:
+        return np.nan
+    raise ValueError("No value is most frequent.")
+
+
+def unique(x: pd.Series) -> Any:
+    """Return single unique value (or error if none exists)."""
+    x = x.dropna()
+    if x.empty:
+        return np.nan
+    uniques = x.unique()
+    if uniques.size == 1:
+        return uniques[0]
+    raise ValueError("Not unique.")
+
+
+def as_dict(x: pd.Series) -> Dict[Any, list]:
+    """Return dictionary of values, listed by index."""
+    result = {}
+    for key, xi in x.dropna().groupby(x.index):
+        result[key] = list(xi)
+    return result
+
+
+def try_aggfunc(  # noqa: C901
+    func: Callable,
+    method: Literal["coerce", "raise", "return", "insert", "append"] = "coerce",
+    error: Union[str, Callable] = None,
+) -> Callable:
+    """
+    Wrap aggregate function in a try-except for error handling.
+
+    Arguments:
+        func: Aggregate function.
+        method: Error handling method (for :class:`ValueError` only).
+
+            - 'coerce': Return `np.nan` instead of the error.
+            - 'raise': Re-raise the error.
+            - 'return' and 'insert': Return rather than raise the error.
+            - 'append': Append the error to a list passed as a second function argument.
+
+        error: Error value, whose type and format depends on `method`.
+            Below, `x` is the original input and `e` is the original error.
+
+            - 'raise': A string with substitions (e.g. 'Error at {x.name}: {e}')
+              By default, the original error is raised.
+            - 'return': A function with signature `f(x, e)` returning a value to return.
+              By default, the original error is returned.
+            - 'insert': Same as for 'return', except the returned value replaces the
+              arguments of the original error and the original error is returned.
+            - 'append': Same as for 'return', except the returned value is appended.
```

Comment:
can you explain the need for all of these options?

Reply:
The diversity of choices comes out of my experimentations with performance.

- `coerce` is used by `ResourceBuilder.build(..., errors='coerce')`
- `raise` is used by `ResourceBuilder.build(..., errors='raise')`
- `insert` is used by `ResourceBuilder.build(..., errors='report')`
- `return` is unused since I found that `append` was faster for `groupby_apply(errors='report')`
- `append` is unused since I found that it was faster to use the same strategy from within `groupby_apply(errors='report')`

Note that `groupby_apply` exists alongside `groupby_aggregate` in case the latter, faster option is found to not be flexible enough for `ResourceBuilder.build()` in the long run.

Reply:
So, let me try to translate this into outcomes...
- `coerce` will return nulls for errors.
- `raise` will fail on errors.
- `insert` will put the error into the column (which will then be pulled out for reporting).
- `return` and `append` should be removed? because you found faster way to do this.

We'll be able to set these by column or by harvest run? In terms of expected use patterns, I assume that within a standard etl run we'd want run the harvesting with `coerce` and do some per-column tests to ensure we are getting enough non-nulls in the harvested columns. And we'd run `insert`/report when that fails and we are trying to debug. I'm sure there are instances when we'd want to run `raise` on specific columns.... but we don't do that right now and I can't think of a specific instance right now.

Also, how are these error args different than the `method` args above? they seem similar and related? For simplicity is it possible to condense them into these three expected behaviors (coerce, raise, insert)?

Also, also, if we are getting rid of `return` and `append`, perhaps we could rename `insert` to be `report` to make it more clear?

Reply:
@cmgosnell I can remove the `return` and `append` options and rename `insert` to `report`.

Setting error behavior is currently only supported by harvest run. Let me know if you would like support for setting it by column.

Reply:
Great! And I think setting this by harvest run will be just fine.